plugins/processors/awsappsignals/internal/resolver/eks.go

```diff
@@ -0,0 +1,739 @@
+	deleter                        Deleter
+}
+
+func NewServiceToWorkloadMapper(serviceAndNamespaceToSelectors, workloadAndNamespaceToLabels, serviceToWorkload *sync.Map, logger *zap.Logger, deleter Deleter) *ServiceToWorkloadMapper {
+	return &ServiceToWorkloadMapper{
+		serviceAndNamespaceToSelectors: serviceAndNamespaceToSelectors,
+		workloadAndNamespaceToLabels:   workloadAndNamespaceToLabels,
+		serviceToWorkload:              serviceToWorkload,
+		logger:                         logger,
+		deleter:                        deleter,
+	}
+}
+
+func (m *ServiceToWorkloadMapper) MapServiceToWorkload() {
+	m.logger.Debug("Map service to workload at:", zap.Time("time", time.Now()))
+
+	m.serviceAndNamespaceToSelectors.Range(func(key, value interface{}) bool {
+		var workloads []string
+		serviceAndNamespace := key.(string)
+		_, serviceNamespace := extractResourceAndNamespace(serviceAndNamespace)
+		serviceLabels := value.(mapset.Set[string])
+
+		m.workloadAndNamespaceToLabels.Range(func(workloadKey, labelsValue interface{}) bool {
+			labels := labelsValue.(mapset.Set[string])
+			workloadAndNamespace := workloadKey.(string)
+			_, workloadNamespace := extractResourceAndNamespace(workloadAndNamespace)
+			if workloadNamespace == serviceNamespace && workloadNamespace != "" && serviceLabels.IsSubset(labels) {
+				m.logger.Debug("Found workload for service", zap.String("service", serviceAndNamespace), zap.String("workload", workloadAndNamespace))
+				workloads = append(workloads, workloadAndNamespace)
+			}
+
+			return true
+		})
+
+		if len(workloads) > 1 {
+			m.logger.Info("Multiple workloads found for service. You will get unexpected results.", zap.String("service", serviceAndNamespace), zap.Strings("workloads", workloads))
+		} else if len(workloads) == 1 {
+			m.serviceToWorkload.Store(serviceAndNamespace, workloads[0])
+		} else {
+			m.logger.Debug("No workload found for service", zap.String("service", serviceAndNamespace))
+			m.deleter.DeleteWithDelay(m.serviceToWorkload, serviceAndNamespace)
+		}
+		return true
+	})
+}
+
+func (m *ServiceToWorkloadMapper) Start(stopCh chan struct{}) {
+	// do the first mapping immediately
+	m.MapServiceToWorkload()
+	m.logger.Debug("First-time map service to workload at:", zap.Time("time", time.Now()))
+
+	go func() {
+		for {
+			select {
+			case <-stopCh:
+				return
+			case <-time.After(time.Minute + 30*time.Second):
+				m.MapServiceToWorkload()
+				m.logger.Debug("Map service to workload at:", zap.Time("time", time.Now()))
+			}
+		}
+	}()
+}
+
+func getEksResolver(logger *zap.Logger) subResolver {
+	once.Do(func() {
+		config, err := clientcmd.BuildConfigFromFlags("", "")
+		if err != nil {
+			logger.Fatal("Failed to create config", zap.Error(err))
+		}
+
+		clientset, err := kubernetes.NewForConfig(config)
+		if err != nil {
+			logger.Fatal("Failed to create eks client", zap.Error(err))
+		}
+
+		// add a time jitter of 10 seconds
+		jitterSleep(10)
```

Comment:
What is the jitter for?

Reply:
not sure, @thpierce / @nanzhenAWS might know?

Reply:
Just to add a jitter to avoid the case while all instances of cloudwatch agent will call the kubernetes control plane at the same time. The time here is just an arbitrary value. We can probably add an TODO for future optimization. But for now, I don't think we want to do extensive testing to figure out a optimal value.